Add Checksum to value log entries

peterstace opened the pull request:
Changes:

* There is now a 32 bit CRC checksum at the end of each value log entry. So the format is now `[header][key][value][crc32]`.
* It's checked during value log replay. If an mismatch is found, the value log is truncated at the end of the last good entry.
* It's also checked during a normal value log read.
* When doing the value log replay, we also truncate the log if a partial entry is found (i.e. we reach EOF while in the middle of an entry).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dgraph-io/badger/148)
<!-- Reviewable:end -->

srh commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[value.go, line 150 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/148#-KqyowsRhj2UbudyXFkZ:-KqyowsRhj2UbudyXFk_:b-5i3mwy) ([raw file](https://github.com/dgraph-io/badger/blob/c13f2441a32288acdc0945b159df8cdf3920ec16/value.go#L150)):*
> ```Go
> 		if err = read(tee, hbuf[:]); err != nil {
> 			if err == io.EOF {
> 				break
> ```

This call to the local function read() could read a few bytes and then return with io.EOF.  The result is that we have to truncate the file.  (And the bug is that we don't.)  We could use io.ReadFull instead, or we could always truncate, to address this.

---

*[value_test.go, line 277 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/148#-Kqyrt_-ZXko1OkU9hJ6:-Kqyrt_-ZXko1OkU9hJ7:b-l2rq6c) ([raw file](https://github.com/dgraph-io/badger/blob/c13f2441a32288acdc0945b159df8cdf3920ec16/value_test.go#L277)):*
> ```Go
> }
> 
> func TestPatialAppendToValueLog(t *testing.T) {
> ```

Partial

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/148)*
<!-- Sent from Reviewable.io -->

peterstace commented:
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[value.go, line 150 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/148#-KqyowsRhj2UbudyXFkZ:-KqzCtOfik8iFjCk9Zph:bfb6y6i) ([raw file](https://github.com/dgraph-io/badger/blob/c13f2441a32288acdc0945b159df8cdf3920ec16/value.go#L150)):*
<details><summary><i>Previously, srh (Sam Hughes) wrote…</i></summary><blockquote>

This call to the local function read() could read a few bytes and then return with io.EOF.  The result is that we have to truncate the file.  (And the bug is that we don't.)  We could use io.ReadFull instead, or we could always truncate, to address this.
</blockquote></details>

Fixed. I just got rid of the `read` function entirely handle the `io.EOF` and `io.UnexpectedEOF` errors appropriately in each case (only different for the header).

---

*[value_test.go, line 277 at r1](https://reviewable.io:443/reviews/dgraph-io/badger/148#-Kqyrt_-ZXko1OkU9hJ6:-KqzCjx2O6G4-_w9nbFU:b-886gsx) ([raw file](https://github.com/dgraph-io/badger/blob/c13f2441a32288acdc0945b159df8cdf3920ec16/value_test.go#L277)):*
<details><summary><i>Previously, srh (Sam Hughes) wrote…</i></summary><blockquote>

Partial
</blockquote></details>

Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/148)*
<!-- Sent from Reviewable.io -->

srh commented:
Everything else looks correct to my eyes.  My main other concern is performance.

The general question is how much should we care -- disks are slow, right?  I'm not so sure about the future though, what with NVM devices.

Going by this:

https://golang.org/src/hash/crc32/crc32_amd64.go

It looks like Golang supports the amd64 CRC32 instruction if we use Castagnoli instead of Koopman.  This would make it about as fast as using Adler32 or Fletcher checksums on that platform.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/148#-:-KqzJ2VIzsmfOYCgrNYL:bpdpcpj)*
<!-- Sent from Reviewable.io -->

peterstace commented:
The whole code path for Castagnoli seems more specialised (short cuts in a few places), so using that instead.

---

Review status: 0 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/148#-:-KqzYUYe-n_Gbc0sHkIM:byccpnc)*
<!-- Sent from Reviewable.io -->

srh commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/148#-:-Kqz_1_Ox_Cwlnz8bLPL:bnfp4nl)*
<!-- Sent from Reviewable.io -->

manishrjain commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[value.go, line 107 at r3](https://reviewable.io:443/reviews/dgraph-io/badger/148#-KqzaRu0P2Klx6Tx3-HS:-KqzaRu0P2Klx6Tx3-HT:b-yemun) ([raw file](https://github.com/dgraph-io/badger/blob/71fa0539b19f6c8418c2dd202177f260ffe7f766/value.go#L107)):*
> ```Go
> 
> var errStop = errors.New("Stop iteration")
> ```

Remove vertical space here.

---

*[value.go, line 194 at r3](https://reviewable.io:443/reviews/dgraph-io/badger/148#-Kqzc9hw9vKDNTimAlFF:-Kqzc9hw9vKDNTimAlFG:b-2jo5yg) ([raw file](https://github.com/dgraph-io/badger/blob/71fa0539b19f6c8418c2dd202177f260ffe7f766/value.go#L194)):*
> ```Go
> 		var vp valuePointer
> 
> 		recordOffset += uint32(hlen + kl + vl + entryHashSize)
> ```

Duplication here. Just make sure that there's no special reasoning why they should be two separate calculations.

---

*[value.go, line 209 at r3](https://reviewable.io:443/reviews/dgraph-io/badger/148#-Kqzci5LMyXm4mbyzhRk:-Kqzci5LMyXm4mbyzhRl:b7z690v) ([raw file](https://github.com/dgraph-io/badger/blob/71fa0539b19f6c8418c2dd202177f260ffe7f766/value.go#L209)):*
> ```Go
> 
> 	if truncate {
> 		f.Lock()
> ```

No need for lock acquisition.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/148#-:-KqzdfXi3OIcgxEw-jFh:bnfp4nl)*
<!-- Sent from Reviewable.io -->

peterstace commented:
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[value.go, line 107 at r3](https://reviewable.io:443/reviews/dgraph-io/badger/148#-KqzaRu0P2Klx6Tx3-HS:-Kqzs4Qt-RKfTZN3KC_H:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/71fa0539b19f6c8418c2dd202177f260ffe7f766/value.go#L107)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Remove vertical space here.
</blockquote></details>

Done.

---

*[value.go, line 194 at r3](https://reviewable.io:443/reviews/dgraph-io/badger/148#-Kqzc9hw9vKDNTimAlFF:-Kqzs6qlA4kgvxfWYgSF:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/71fa0539b19f6c8418c2dd202177f260ffe7f766/value.go#L194)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

Duplication here. Just make sure that there's no special reasoning why they should be two separate calculations.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/148)*
<!-- Sent from Reviewable.io -->

peterstace commented:
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[value.go, line 209 at r3](https://reviewable.io:443/reviews/dgraph-io/badger/148#-Kqzci5LMyXm4mbyzhRk:-Kqzu2sWAGQhpW2osIiP:b-896fix) ([raw file](https://github.com/dgraph-io/badger/blob/71fa0539b19f6c8418c2dd202177f260ffe7f766/value.go#L209)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary><blockquote>

No need for lock acquisition.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/badger/148)*
<!-- Sent from Reviewable.io -->
